transport/grpc/stream.go

```diff
@@ -0,0 +1,219 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package grpc
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"io/ioutil"
+
+	"github.com/opentracing/opentracing-go"
+	"go.uber.org/atomic"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/yarpcerrors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+type serverStream struct {
+	ctx    context.Context
+	req    *transport.StreamRequest
+	stream grpc.ServerStream
+}
+
+func newServerStream(ctx context.Context, req *transport.StreamRequest, stream grpc.ServerStream) *serverStream {
+	return &serverStream{
+		ctx:    ctx,
+		req:    req,
+		stream: stream,
+	}
+}
+
+func (ss *serverStream) Context() context.Context {
+	return ss.ctx
+}
+
+func (ss *serverStream) Request() *transport.StreamRequest {
+	return ss.req
+}
+
+func (ss *serverStream) SendMessage(_ context.Context, m *transport.StreamMessage) error {
+	// TODO pool buffers for performance.
+	msg, err := ioutil.ReadAll(m.Body)
+	_ = m.Body.Close()
+	if err != nil {
+		return err
+	}
+	return toYARPCStreamError(ss.stream.SendMsg(msg))
+}
+
+func (ss *serverStream) ReceiveMessage(_ context.Context) (*transport.StreamMessage, error) {
+	// TODO used pooled buffers for performance.
+	var msg []byte
+	if err := ss.stream.RecvMsg(&msg); err != nil {
+		s, ok := status.FromError(err)
+		if ok && s.Code() == codes.Canceled && s.Message() == context.Canceled.Error() {
```

Comment:
Is the string match here absolutely necessary? If yes, can you please clarify
that in the comment below in more detail, preferably linking to the relevant
GRPC Go issue.

Reply:
whoops, double checked and that was actually invalid.  Removing the check here, and I'll modify the failing test to avoid the race condition.